src/Raven.Server/Commercial/LetsEncryptUtils.cs

```diff
@@ -0,0 +1,968 @@
+                }
+                catch (Exception e)
+                {
+                    if (cts.IsCancellationRequested == false)
+                        throw;
+                    throw new TimeoutException("Request failed due to a timeout error", e);
+                }
+            }
+        }
+
+        public static async Task<(string Challenge, LetsEncryptClient.CachedCertificateResult Cache)> InitialLetsEncryptChallenge(
+            SetupInfo setupInfo,
+            LetsEncryptClient client,
+            CancellationToken token)
+        {
+            try
+            {
+                var host = (setupInfo.Domain + "." + setupInfo.RootDomain).ToLowerInvariant();
+                var wildcardHost = "*." + host;
+                if (client.TryGetCachedCertificate(wildcardHost, out var certBytes))
+                    return (null, certBytes);
+
+                var result = await client.NewOrder(new[] {wildcardHost}, token);
+
+                result.TryGetValue(host, out var challenge);
+                // we may already be authorized for this?
+                return (challenge, null);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Failed to receive challenge(s) information from Let's Encrypt.", e);
+            }
+        }
+
+        public class CompleteAuthorizationAndGetCertificateParameters
+        {
+            public Action OnValidationSuccessful;
+            public SetupProgressAndResult Progress;
+            public SetupInfo SetupInfo;
+            public LetsEncryptClient Client;
+            public (string Challange, LetsEncryptClient.CachedCertificateResult Cache) ChallengeResult;
+            public ServerStore ServerStore;
+            public CancellationToken Token;
+        }
+
+        public class CompleteClusterConfigurationParameters
+        {
+            public SetupProgressAndResult Progress;
+            public Action<IOperationProgress> OnProgress;
+            public SetupInfo SetupInfo;
+            public Func<string,string, Task> OnBeforeAddingNodesToCluster;
+            public Func<string, Task> AddNodeToCluster;
+            public Func<Action<IOperationProgress>, SetupProgressAndResult, X509Certificate2, Task> RegisterClientCertInOs;
+            public Func<StudioConfiguration.StudioEnvironment, Task> OnPutServerWideStudioConfigurationValues;
+            public Func<string, Task> OnWriteSettingsJsonLocally;
+            public Func<string, Task<string>> OnGetCertificatePath;
+            public Func<string, Task> RegisterClientCert;
+            public Func<CertificateDefinition, Task> PutCertificate;
+            public SetupMode SetupMode;
+            public string SettingsPath;
+            public bool CertificateValidationKeyUsages;
+            public LicenseType LicenseType;
+            public CancellationToken Token = CancellationToken.None;
+        }
+
+        public class CertificateHolder
```

Comment:
Don't we have such class already?

Reply:
We had in `SetupManager` and I moved it to `LetsEncryptUtils`

Reply:
Not exactly. You moved it from `RavenServer.cs` as I see. So there is no duplicate - that's great.  But now you have the code like:

```
private LetsEncryptUtils.CertificateHolder LoadCertificateAtStartup()
```

which suggest it's Let's Encrypt specific certificate. That's not always the case because self-signed certificate will also be loaded as startup and returned here. That introduces a bit of unnecessary confusion when reading the code.

I think `CertificateHolder ` might be just a standalone class that you will be able to link in `rvn` project